Always allow full filesystem access to LSP.

chriseth opened the pull request:
Maybe we need a different solution for Windows? Something like a flag to just disable the check inside the file reader?

ekpyron commented:
@christianparpart For CI to pass this will need a rebase to current ``develop``.

chriseth commented on libsolidity/lsp/FileRepository.h:
```diff
@@ -29,7 +29,10 @@ class FileRepository
-		m_fileReader(_basePath) {}
+		m_fileReader(_basePath)
+	{
+		m_fileReader.allowDirectory(boost::filesystem::path("/"));
```
Does this work on windows?

christianparpart replied:
Yes, with my changes added to this PR: All tests pass (locally).

christianparpart committed:
[Circle CI] Adds LSP tests to Windows CI.

chriseth commented on test/lsp.py:
```diff
@@ -180,6 +177,10 @@ def __init__(self):
         args = create_cli_parser().parse_args()
         self.solc_path = args.solc_path
         self.project_root_dir = os.path.realpath(args.project_root_dir) + "/test/libsolidity/lsp"
+        if os.name == 'nt':
```
What about using `as_uri()` from `pathlib`?

christianparpart replied:
I did not know about that. will be trying.

chriseth commented on test/lsp.py:
```diff
@@ -60,10 +60,7 @@ def receive_message(self) -> Union[None, dict]:
                 # server quit
                 return None
             line = line.decode("utf-8")
-            if not line.endswith("\r\n"):
```
Why did you remove this check?

christianparpart replied:
the rstrip does also remove CRLF but more relaxed, such that it would have allowed a simple LF too. I've reverted it though.

chriseth commented on libsolidity/lsp/Transport.cpp:
```diff
@@ -107,8 +107,15 @@ void IOStreamTransport::send(Json::Value _json, MessageID _id)
 
-	m_output << "Content-Length: " << jsonString.size() << "\r\n";
-	m_output << "\r\n";
+#if defined(_WIN32)
```
I don't think this is a proper solution. I saw some posts mentioning freopen and others `setmode`: https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/setmode?view=msvc-170&viewFallbackFrom=vs-2019

chriseth commented on libsolidity/lsp/FileRepository.cpp:
```diff
@@ -42,7 +42,16 @@ string FileRepository::sourceUnitNameToClientPath(string const& _sourceUnitName)
+#if defined(_WIN32)
+		// Ensure basePath does contain Windows drive letter to fully qualify the path.
+		if (!basePath.has_root_name())
+			basePath = (boost::filesystem::current_path().root_path() / basePath).normalize();
```
I don't think we should use the current path.

chriseth commented on libsolidity/lsp/Transport.cpp:
```diff
@@ -75,67 +70,148 @@ optional<Json::Value> IOStreamTransport::receive()
+	auto const n = fread(buffer.data(), 1, _byteCount, stdin);
+	if (n < _byteCount)
+		buffer.resize(n);
+	lspDebug(fmt::format("StdioTransport.readBytes({};{};{}): \"{}\"",
```
coding style

chriseth commented on libsolidity/lsp/Transport.cpp:
```diff
@@ -75,67 +70,148 @@ optional<Json::Value> IOStreamTransport::receive()
+{
+	char* lineBuffer = nullptr;
+	size_t numBytesRead = 0;
+	auto const getlineResult = ::getline(&lineBuffer, &numBytesRead, stdin);
```
Can you use an explicit type please?

chriseth commented on libsolidity/lsp/Transport.cpp:
```diff
@@ -75,67 +70,148 @@ optional<Json::Value> IOStreamTransport::receive()
+	char* lineBuffer = nullptr;
+	size_t numBytesRead = 0;
+	auto const getlineResult = ::getline(&lineBuffer, &numBytesRead, stdin);
+	auto const ensureFreedBuffer = ScopeGuard{ [&]() { free(lineBuffer); }};
```
```suggestion
	ScopeGuard const ensureFreedBuffer{[&]() { free(lineBuffer); }};
```

chriseth commented on libsolidity/lsp/Transport.cpp:
```diff
@@ -75,67 +70,148 @@ optional<Json::Value> IOStreamTransport::receive()
+
+void StdioTransport::writeBytes(std::string_view _data)
+{
+	auto const bytesWritten = fwrite(_data.data(), 1, _data.size(), stdout);
```
Please use an explicit type

leonardoalt commented:
Is this PR waiting for fixes or reviews? @chriseth @christianparpart

christianparpart commented:
> Is this PR waiting for fixes or reviews? @chriseth @christianparpart

yes, please. :)

christianparpart committed:
Fixes to lsp.py with respect to Windows drive-letter paths in URI.

christianparpart committed:
Always allow full filesystem access to LSP.

christianparpart committed:
lsp.py: Port to support running on Windows & adapt to changes due to prior merged PR.

- lsp.py: Fixes invalid-syntax by Python interpreter on Windows CI (older Python version).
- lsp.py: Savely strip CRLF from right side of the string, ignoring accidental multiple occurrences of \r (such as \r\r\n).
- lsp.py: Fixes reading single character from stdin (wrt. Windows platform).
- lsp.py: Adds header line reading to I/O tracing (useful for debugging).
- lsp.py: When running the tests on Windows, don't care test file content's newlines but simply expect LFs (instead of CRLF for example).
- Apply pylint notes.
- Fixing use of @functools.lru_cache for older python versions (CircleCI Windows)

Marenz commented on libsolidity/lsp/FileRepository.cpp:
```diff
@@ -17,48 +17,129 @@
+
+		string const strippedSourceUnitName = stripFileUriSchemePrefix(_sourceUnitName);
+
+		if (boost::filesystem::path(strippedSourceUnitName).has_root_path() &&
```
`if (` should be on its own

Marenz reviewed (changes requested):
Found a tiny styling problem, but pylint also demands fixes.

Marenz commented on libsolidity/lsp/Transport.cpp:
```diff
@@ -75,79 +69,156 @@ optional<Json::Value> IOStreamTransport::receive()
+
+		auto const name = boost::to_lower_copy(line.substr(0, delimiterPos));
+		auto const value = line.substr(delimiterPos + 1);
+		if (!headers.emplace(
```
`if (` should be on its own

christianparpart replied:
This is just moved code, not new one. I change change however. :)

Marenz commented on test/lsp.py:
```diff
@@ -1,26 +1,57 @@
+    else:
+        # Turn off user input buffering so we get the input immediately,
+        # not only after a line break
+        tty.setcbreak(sys.stdin.fileno())
```
This doesn't need to be set every time we want a char. Once in the life-time of the program is enough.

christianparpart replied:
Exactly, but that way I kept the logic more closely together without over-complifying something simple.

chriseth commented on libsolidity/lsp/Transport.h:
```diff
@@ -119,27 +143,34 @@ class IOStreamTransport: public Transport
 };
 
+/**
+ * Standard I/O transport Layer utilizing stdin/stdout for communication.
```
The changes in these files do not look like they are about allowing full filesystem access. Please put them in their own PR in the future.

christianparpart replied:
I was considering that. Okay.
